### PR TITLE
feat: init cluster manager framework

### PR DESCRIPTION
**Pull request by wangyizhi1**
#### What type of PR is this?

feat

#### What does this PR do?

init cluster manager framework

### CODE REVIEW

File: cmd/clustertree/cluster-manager/app/manager.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	if err = ClusterController.SetupWithManager(mgr); err != nil {
+		return fmt.Errorf("error starting %s: %v", clusterManager.ControllerName, err)
+	}
+
+	if err := mgr.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start controller manager: %v", err)
+	}
+
+	return nil
```
Comment (xyz2277) [inline cmd/clustertree/cluster-manager/app/manager.go:80]: it seems that you lost LeaderElection

**Comment by duanmengkk** [comment]
lgtm

File: pkg/clustertree/cluster-manager/cluster_controller.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+		Named(ControllerName).
+		WithOptions(controller.Options{}).
+		For(&clusterlinkv1alpha1.Cluster{}, builder.WithPredicates(predicatesFunc)).
+		Complete(c)
+}
+
+func (c *ClusterController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(4).Infof("============ %s starts to reconcile %s ============", ControllerName, request.Name)
+	defer func() {
+		klog.V(4).Infof("============ %s has been reconciled =============", request.Name)
```
Comment (wuyingjun-lucky) [inline pkg/clustertree/cluster-manager/cluster_controller.go:110]: I think it is not suitable to place a defer function for logging that will make a mis understanding

File: pkg/clustertree/cluster-manager/cluster_controller.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	c.ControllerManagers[cluster.Name] = &mgr
+	c.ManagerCancelFuncs[cluster.Name] = &cancel
+	c.ControllerManagersLock.Unlock()
+
+	if err = c.setupControllers(&mgr); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to setup cluster %s controllers: %v", cluster.Name, err)
+	}
+
+	go func() {
+		if err := mgr.Start(subContext); err != nil {
```
Comment (wuyingjun-lucky) [inline pkg/clustertree/cluster-manager/cluster_controller.go:178]: controller on controller ？ what does it mean ？

Comment (wangyizhi1) [reply pkg/clustertree/cluster-manager/cluster_controller.go:178]: Yes, when the cluster joins, start the controller manager of the cluster

File: pkg/clustertree/cluster-manager/cluster_controller.go (hunk 2 above)
Comment (wangyizhi1) [reply pkg/clustertree/cluster-manager/cluster_controller.go:110]: This is not a big problem, I will optimize it next time

File: pkg/clustertree/cluster-manager/cluster_controller.go (hunk 3 above)
Comment (wuyingjun-lucky) [reply pkg/clustertree/cluster-manager/cluster_controller.go:178]: > Yes, when the cluster joins, start the controller manager of the cluster

One member cluster will have a single controller manager?

Comment (wangyizhi1) [reply pkg/clustertree/cluster-manager/cluster_controller.go:178]: yes

**Pull request merged**